### DISCUSSION TITLE
pvpython error in trame

### DISCUSSION THREAD

**Discussion by earlsioson** [Kitware/trame#289] [Q&A] [answered]
Hello,

We are getting the following error in our trame app...

ERROR:root:Exception raised by task = <Task finished name='Task-12' coro=<main.<locals>.reload() done, defined at /workspace/trame/visualization/visualization/app/main.py:49> exception=ImportError("Can't connect to HTTPS URL because the SSL module is not available.")>

We are using python 3.9.5 and running trame with the pvpython command. We started receiving the error after we added functionality to download files from Google cloud storage. 

We verified that our download functions work correctly outside of trame on the same VM using normal python command so we think the issue is with pvpython.

Any help or guidance is appreciated.

**Comment by jourdain** [comment, marked as answer]
Indeed, the issue is related to the fact that with ParaView downloaded binaries we do not bundle SSL to prevent security issue. 

In your case, you will have to build paraview with your system python so ParaView can have access to your system SSL. 
If you are using docker to bundle your app, you can do something along those lines.

```docker
# Install packages needed for building ParaView
RUN apt-get update -y \
    && apt-get install -y \
        # Build PV
        cmake build-essential libgl1-mesa-dev libxt-dev ninja-build python3 python3-dev python3-numpy freeglut3-dev \
    # Clean
    && rm -rf /var/lib/apt/lists/*

# Build ParaView with EGL and system python
RUN mkdir -p /opt/pv-dev/src \
    && cd /opt/pv-dev/src \
    && wget -qO- https://www.paraview.org/files/v5.11/ParaView-v5.11.1.tar.gz | tar --strip-components=1 -xzv

RUN cmake \
        -S /opt/pv-dev/src \
        -B /opt/pv-dev/build \
        -G Ninja \
        -D PARAVIEW_USE_PYTHON=ON \
        -D PARAVIEW_USE_MPI=OFF \
        -D VTK_OPENGL_HAS_EGL=ON \
        -D VTK_USE_X=OFF \
        -D VTK_DEFAULT_RENDER_WINDOW_HEADLESS=ON \
        -D CMAKE_BUILD_TYPE=Release \
        -D PARAVIEW_USE_QT=OFF \
    && cmake --build /opt/pv-dev/build \
    && cmake --install /opt/pv-dev/build --prefix /opt/paraview

ENV TRAME_PARAVIEW=/opt/paraview
```

Sometime we trigger the CI with SSL ON and download those binaries for those use-cases. 

HTH

**Reply by earlsioson** [reply]
Thank you @jourdain, I will give what you suggest a try.

**Reply by earlsioson** [reply]
@jourdain I'm stuck with this error...
-- Could NOT find Python3 (missing: Development.Module) (found suitable version "3.9.5", minimum required is "3.4")
CMake Error at VTK/CMake/vtkModule.cmake:4579 (message):
  Could not find the Python3 external dependency.
Call Stack (most recent call first):
  VTK/Utilities/Python/CMakeLists.txt:55 (vtk_module_find_package)

Do you know how I can install the missing: Development.Module?

**Reply by jourdain** [reply]
In which context, linux, mac, docker?

`python3-dev` is supposed to provide such thing. But without any information it is impossible to tell.

**Reply by earlsioson** [reply]
@jourdain, It's in a docker container using kitware/trame:glvnd as the base image and after running those command you listed above.

**Reply by earlsioson** [reply]
@jourdain , nevermind, I had to explicitly install python3.9-dev, python3-dev was installing python3.8-dev. After compiling I was able to down via ssl. Thanks again for your assistance.
